streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -788,9 +793,9 @@ protected StreamTask createStreamTask(TaskId id, Collection<TopicPartition> part
 
         sensors.taskCreationSensor.record();
 
-        ProcessorTopology topology = builder.build(id.topicGroupId);
-
-        return new StreamTask(id, applicationId, partitions, topology, consumer, producer, restoreConsumer, config, sensors, stateDirectory, cache);
+        final ProcessorTopology topology = builder.build(id.topicGroupId);
+        final RecordCollector recordCollector = new RecordCollectorImpl(producer, id.toString());
+        return new StreamTask(id, applicationId, partitions, topology, consumer, restoreConsumer, config, sensors, stateDirectory, cache, recordCollector);
```

Comment:
The producer is only used to construct the `RecordCollectorImpl`. Design wise it is better to have `StreamTask` constructed with its dependency on `RecordCollector` being obvious. It also promotes less coupling as `RecordCollector` is now an interface, and makes the code easier to test. 
As it stands, a lot of this code is difficult to change, and to test,  due to high-coupling. This should change.